### ISSUE TITLE
should int to uint implicit conversion add a runtime check?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#20543]
This issue is about adding runtime checking for certain implicit conversions from int to uint. Issue  #20091 discusses adding such implicit conversions and PR #20528 adds them.

After PR #20528, we can write code that changes a negative `int` to to a `uint` implicitly. For example:

``` chapel
proc f(arg: uint) { writeln("in f, arg=", arg); }

f(-1);
```

produces this output:
```
in f, arg=18446744073709551615
```

Should we seek to add some checking that can be enabled with `--checks` that would halt if a negative number is implicitly converted in this way?

**Comment by bradcray** [comment]
On one hand, that seems like it would be attractive.  On the other hand, it seems similar to other such "value range" checks that we've talked about adding over time, so I'm not sure if there's a strong rationale for adding this one without adding the others, or before doing so.  I suppose that one might be that it's new and may be more surprising.

**Comment by damianmoz** [comment]
Sometimes there is a valid reason for doing such as assignment so halting is a bit anti-social.

However, a compiler warning is quite in order is such an assignment is seen.

If I am worried about such an assignment., I stick a judiciously placed **assert** statement in there.

My 2c.

**Comment by mppf** [comment]
@damianmoz - one more thought on your post

> Sometimes there is a valid reason for doing such as assignment so halting is a bit anti-social.

Right, but as with the warnings, one can avoid the warning or halt by adding a cast, which arguably makes the code clearer. E.g. `~ (myInt: uint)`.

**Comment by damianmoz** [comment]
I did not know about `--warn-int-uint`. Is that new?  I cannot find documentation on it. Anyway, it looks perfect. My personal opinion is that warnings be on by default and that the user should be forced to turn them off. But I am not overly dogmatic on that.

I will ponder your second comment for now before replying.

**Comment by mppf** [comment]
> I did not know about `--warn-int-uint`. Is that new? I cannot find documentation on it. Anyway, it looks perfect. My personal opinion is that warnings be on by default and that the user should be forced to turn them off. But I am not overly dogmatic on that.

Yes, it is new (as of last week) and it is currently a Developer flag which means it isn't really documented yet. If we conclude that "Issue a warning by default and a provide a way to turn the warning off" is the strategy we want here, we will make it user-facing in the next release. But for now it will be available in the 1.28 release to experiment with.

**Comment by mppf** [comment]
I added Chapel 2.0 to this issue just because I think the question of "what is the default" should be settled for 2.0. My current leaning is that the existing int->uint compilation warning should be on by default and that we shouldn't add runtime checking at this time. And, we can add features to enable ways to opt-in to more warnings (but those wouldn't be a breaking change).

**Comment by damianmoz** [comment]
100$ behind those sentiments of @mppf

**Comment by bradcray** [comment]
I agree with not having a runtime check on by default, as it doesn't seem inherently so different than other cases where one might want a "value check" during a conversion (something we've discussed supporting for years in an opt-in basis, but haven't added yet—I don't feel inclined to add it for this case ahead of others, and wouldn't make it default-on even if we did due to performance overheads).

I feel nervous about having the compilation warning on by default for similar reasons (why this case and not others?) and because I'm not sure how much noise it will cause in practice.  But I'm open to giving it a try (or for waiting for the first incensed user to show up because we turned their negative int into a large uint).

**Comment by mppf** [comment]
@bradcray - do you think it would be acceptable to turn the warning on by default after Chapel 2.0? If so, I think we should remove the Chapel 2.0 label from this issue & table this issue for the time being.

**Comment by bradcray** [comment]
> do you think it would be acceptable to turn the warning on by default after Chapel 2.0?

Yeah, I think so.  We're not changing the behavior of the code after all by doing that, just the compiler.  And presumably the compiler will only get better and better w.r.t. warnings as time goes on (where "better" should not necessarily be interpreted as "warn about more and more things by default").

**Comment by mppf** [comment]
Since the compiler currently warns about implicit conversion from signed to unsigned with `--warn-unstable`, turning it on or off or even making it an error after Chapel 2.0 should be OK. I originally added the "Chapel 2.0" label to this issue to indicate that it would be nice if we can conclude this before then; however as that is a "nice to have" thing, I'm going to remove the "Chapel 2.0" label here.
